python/dgl/sparse.py

```diff
@@ -366,124 +366,75 @@ def _bwd_segment_cmp(feat, arg, m):
                                  to_dgl_nd_for_write(out))
     return out
 
-class CSRMatrix(object):
-    """Device- and backend-agnostic sparse matrix in CSR format.
+def csrmm(A, A_weights, B, B_weights, num_vtypes):
+    """Return a graph whose adjacency matrix is the sparse matrix multiplication
+    of those of two given graphs.
 
     Parameters
     ----------
-    data : Tensor
-        The data array.
-    indices : Tensor
-        The column indices array.
-    indptr : Tensor
-        The row index pointer array.
-    num_rows : int
-        The number of rows.
-    num_cols : int
-        The number of columns.
-    """
-    def __init__(self, data, indices, indptr, num_rows, num_cols):
-        self.indptr = indptr
-        self.indices = indices
-        self.data = data
-        self.shape = (num_rows, num_cols)
-
-def csrmm(A, B):
-    """Sparse-sparse matrix multiplication.
-
-    This is an internal function whose interface is subject to changes.
-
-    Parameters
-    ----------
-    A : dgl.sparse.CSRMatrix
-        The left operand
-    B : dgl.sparse.CSRMatrix
-        The right operand
+    A : HeteroGraphIndex
+        The input graph index as left operand.
+    A_weights : Tensor
+        The edge weights of graph A as 1D tensor.
+    B : HeteroGraphIndex
+        The input graph index as right operand.
+    B_weights : Tensor
+        The edge weights of graph B as 1D tensor.
+    num_vtypes : int
+        The number of node types for the returned graph (must be either 1 or 2).
 
     Returns
     -------
-    dgl.sparse.CSRMatrix
-        The result
+    C : HeteroGraphIndex
+        The output graph index.
+    C_weights : Tensor
+        The edge weights of the output graph.
     """
-    A_indptr = F.zerocopy_from_numpy(A.indptr)
-    A_indices = F.zerocopy_from_numpy(A.indices)
-    A_data = F.zerocopy_from_numpy(A.data)
-    B_indptr = F.zerocopy_from_numpy(B.indptr)
-    B_indices = F.zerocopy_from_numpy(B.indices)
-    B_data = F.zerocopy_from_numpy(B.data)
-    C_indptr, C_indices, C_data = _CAPI_DGLCSRMM(
-        A.shape[0], A.shape[1], B.shape[1],
-        F.to_dgl_nd(A_indptr),
-        F.to_dgl_nd(A_indices),
-        F.to_dgl_nd(A_data),
-        F.to_dgl_nd(B_indptr),
-        F.to_dgl_nd(B_indices),
-        F.to_dgl_nd(B_data))
-    return CSRMatrix(
-        F.from_dgl_nd(C_data),
-        F.from_dgl_nd(C_indices),
-        F.from_dgl_nd(C_indptr),
-        A.shape[0],
-        B.shape[1])
-
-def csrsum(As):
-    """Sparse-sparse matrix summation.
-
-    This is an internal function whose interface is subject to changes.
+    C, C_weights = _CAPI_DGLCSRMM(
+        A, F.to_dgl_nd(A_weights), B, F.to_dgl_nd(B_weights), num_vtypes)
+    return C, F.from_dgl_nd(C_weights)
+
+def csrsum(As, A_weights):
```

Comment:
> Does it mean that the graph must be stored in a CSR format? What if the format is restricted to COO?
It must be in CSR.  cuSPARSE does not support other formats.
> I think "graph_union" is actually a better name.
I don't think so.  This case is more restricted because all graphs must have the same set of nodes.

This function is also for internal use and the user interface will be "graph_sum" or something.

Reply:
> It must be in CSR. cuSPARSE does not support other formats.

But the argument type is `GraphIndex`, which causes confusion.

Reply:
Then how about we explicitly require the users to enable CSRs in the user interface?  I don't think we want to write COO kernels.

Reply:
Perhaps it's better to discuss this in the PR for user-facing APIs. Let's just take a note here.